### DISCUSSION TITLE
Config files for mappings

### DISCUSSION THREAD

**Discussion by thijstriemstra** [corrados/edrumulus#19] [Ideas]
Currently pad mappings are hardcoded in `edrumulus.ino` but it would be nice to define this in a separate config file.

So this part, for example, would ideally be defined and loaded from a config file:

```c++
  // some fundamental settings which do not change during operation
  edrumulus.set_midi_notes      ( 0, 38, 40 ); // snare
  edrumulus.set_midi_notes      ( 1, 36, 36 ); // kick
  edrumulus.set_midi_notes      ( hihat_pad_idx, 22 /*42*/, 22 );
  edrumulus.set_midi_notes_open ( hihat_pad_idx, 26 /*46*/, 26 );
  edrumulus.set_midi_notes      ( hihatctrl_pad_idx, 44, 44 ); // Hi-Hat pedal hit
  edrumulus.set_midi_ctrl_ch    ( hihatctrl_pad_idx, 4 ); // Hi-Hat control
  edrumulus.set_midi_notes      ( 4, 49, 55 ); // crash
  edrumulus.set_midi_notes      ( 5, 48, 50 ); // tom 1
  edrumulus.set_midi_notes      ( 6, 51, 53 /*59*/ ); // ride (edge: 59, bell: 53)
  edrumulus.set_midi_notes      ( 7, 45, 47 ); // tom 2
  edrumulus.set_midi_notes      ( 8, 43, 58 ); // tom 3
```

I would prefer a YAML style config file, but I don't know if libyaml is something you can import/include in a Arduino project. Since this projects targets high performance microcontrollers I would hope libyaml fits in memory etc.

Thoughts?

**Comment by corrados** [comment]
These settings like the MIDI notes, pad type and other settings are changed with MIDI commands. If you run Edrumulus from Linux, I have written a startup script which can include all your custom settings. See: https://github.com/corrados/edrumulus/blob/a3d9a1ed934c62814bdced4c13b1c44f99e6eb20/tools/run_edrumulus.sh#L87
There is also another possibility to define your custom pad settings which also works under Windows. You have to install Octave and run [this script](https://github.com/corrados/edrumulus/blob/main/tools/edrumuluscontrol.m). Then a GUI like this shows up:
![grafik](https://user-images.githubusercontent.com/46655886/145913517-8e4790b9-1bd4-4d79-9454-d7a7991cd11d.png)
There is a button Default Settings in that dialog which assigns all your custom pad settings in one step to Edrumulus. You just have to modify the function in the Octave script with your custom settings: https://github.com/corrados/edrumulus/blob/a3d9a1ed934c62814bdced4c13b1c44f99e6eb20/tools/edrumuluscontrol.m#L513

**Comment by thijstriemstra** [comment]
This type of commit is the direction I am looking for: https://github.com/corrados/edrumulus/commit/a54f5d7a31bdbf41c1470444b286c28fe4d74156

Eventually, (re)loading these 'config' files during runtime would be my goal here.
